### PR TITLE
Change avro schema generated for EmailFeedback

### PR DESCRIPTION
**Pull request by laurence-bird**

### CODE REVIEW

File: modules/core/src/test/scala/com/ovoenergy/comms/model/FeedbackSpec.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                           |          "type" : "long",
+                           |          "logicalType" : "timestamp-millis"
+                           |        },
+                           |        "doc" : "A timestamp denoting when the event was created. Value is milliseconds since Unix Epoch - UTC time."
+                           |      } ]
+                           |    }
+                           |  } ]
+                           |}""".stripMargin
+
+    val schemaFor = implicitly[SchemaFor[Feedback]]
```
Comment (SystemFw) [inline modules/core/src/test/scala/com/ovoenergy/comms/model/FeedbackSpec.scala]: perhaps a comment here as to why `SchemaFor[Feedback]` doesn't work (triggers the macro but we need the implicit search since we have overridden the instance)

File: modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala [hunk 2]
```diff
@@ -1,14 +1,82 @@
...
+import com.ovoenergy.comms.model.email.EmailFeedbackOptions.Interacted
+import com.ovoenergy.comms.model.email.EmailFeedbackOptions.Opened.Opened
+import org.apache.avro.generic.{GenericData, GenericRecordBuilder}
+
 sealed trait EmailFeedback
 
+object EmailFeedback {
+
+  implicit val toRecord: ToRecord[EmailFeedback] = new ToRecord[EmailFeedback] {
+    val schemaFor = SchemaFor[Feedback]
```
Comment (filosganga) [inline modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: SchemaFor should be for EmailFeedback

File: modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala [hunk 3]
```diff
@@ -1,14 +1,82 @@
...
+    def apply(t: EmailFeedback): org.apache.avro.generic.GenericRecord =
+      new org.apache.avro.generic.GenericData.Record(schemaFor.apply())
+  }
+
+  implicit def fromValue: FromValue[EmailFeedback] = new FromValue[EmailFeedback] {
+    override def apply(value: Any, field: Schema.Field): EmailFeedback = {
+      val record = value.asInstanceOf[GenericData.Record]
+      Option(record.get(EmailFeedbackOptions.Opened.DummyFieldName))
```
Comment (filosganga) [inline modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: Is it necessary? The record should have schema and name as discriminator

File: modules/core/src/test/scala/com/ovoenergy/comms/model/FeedbackSpec.scala [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+import java.nio.charset.Charset
+
+import com.ovoenergy.comms.model.email.EmailFeedbackOptions
+import com.ovoenergy.comms.model.email.EmailFeedbackOptions.Opened
+import org.scalatest.{EitherValues, FlatSpec, Matchers, OptionValues}
+import io.circe.parser._
+import io.circe._
+import org.scalacheck.Arbitrary
+
+class FeedbackSpec extends FlatSpec with EitherValues with OptionValues with Matchers with Arbitraries {
```
Comment (filosganga) [inline modules/core/src/test/scala/com/ovoenergy/comms/model/FeedbackSpec.scala]: Cool

File: modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala (hunk 3 above)
Comment (laurence-bird) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: hmm as far as I'm aware it is, in the instances of testing any values in field were null and we were unable to get it to pass in any other way. nasty i know D:

Comment (filosganga) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: Take a look here, I believe the `ToSchema` is ignored as well as `ToValue`, `FromValue`:

https://github.com/filosganga/avro-playground/blob/master/src/main/scala/com/github/filosganga/avro/playground/AvroCodec.scala

Keep in mind that `SchemaFor[A]()` does not care if you have an implicit `SchemaFor` in scope, it will generate a new one. The same consideration is valid for `ToRecord`, `FromRecord` and the other avro4s type classes. This lib is really a piece of crap.

Comment (SystemFw) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: We have dealt with this. Not sure why the tests are failing now but we have verified that things are being picked up and work as expected in the repl. I can explain in detail what things are picked or not picked if you want (the library is inconsistent)

Comment (SystemFw) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: We have solved all these problems individually. There is a chance later refactors have messed up some of that when put all together :)

Comment (filosganga) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: Having looked in deep today there are two sets of type classes:

SchemaFor, ToRecord and FromRecord: these are the main typeclasses used to handle root types.

ToSchema, ToValue, FromValue: these are used by the SchemaFor, ToRecord and FromRecord to handle each field of the root type. 

The library can derive these from the formers so you should always implement the former unless you want to override as field is serialized.

Also this mean that despite avrò support it you must have a Record as root type. This is why you cannot test directly EmailFeedback as you want it to be an union.

In short IMHO you should override the SchemaFor and Co for opened and interacted while the ToSchema and Co for EmailFeedback field.

Comment (SystemFw) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: I want to quickly get back to a state where tests are passing, and tomorrow we can look at refactoring it/changing it.

Comment (SystemFw) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: I went to read the library code top to bottom (which I should have done before, in hindsight) and I think you're right. 
We are short-circuiting the typelevel recursion with that `FromValue[EmailFeedback]`, but it we delete it will search for `FromRecord[EmailFeedback]` (but only when it goes through the macro! not with normal implicit resolution), so we can provide that and avoid parsing dummy.

Comment (filosganga) [reply modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: Makes completely sense to me! 🎉

File: modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala [hunk 5]
```diff
@@ -39,36 +57,33 @@ object EmailFeedback {
...
+    }
+
+    implicit def fromRecord = new FromRecord[Opened] {
+      override def apply(record: GenericRecord): Opened = Opened
     }
 
-    val toValue = new ToValue[Opened] {
-      override def apply(t: Opened) = {
-        new GenericRecordBuilder(OpenedSchema)
+    implicit val toRecord = new ToRecord[Opened] {
```
Comment (SystemFw) [inline modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: type signature missing

File: modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala [hunk 6]
```diff
@@ -1,33 +1,51 @@
...
 
-  implicit def fromValue: FromValue[EmailFeedback] = new FromValue[EmailFeedback] {
-    override def apply(value: Any, field: Schema.Field): EmailFeedback = {
-      val record = value.asInstanceOf[GenericData.Record]
-      Option(record.get(EmailFeedbackOptions.Opened.DummyFieldName))
-        .map(_ => EmailFeedbackOptions.Opened)
-        .getOrElse {
-          val urlStr = record.get("url").asInstanceOf[org.apache.avro.util.Utf8].toString
-          Interacted(urlStr)
+  implicit val toSchema = new ToSchema[EmailFeedback] {
```
Comment (SystemFw) [inline modules/core/src/main/scala/com/ovoenergy/comms/model/email/EmailFeedback.scala]: type signature here

File: modules/core/src/main/scala/com/ovoenergy/comms/model/Feedback.scala [hunk 7]
```diff
@@ -25,3 +25,8 @@ case class Feedback(
                   "status" -> status.toString
                 ))
 }
+
+object Feedback {
+
+  implicit def schemaFor = SchemaFor[Feedback]
```
Comment (SystemFw) [inline modules/core/src/main/scala/com/ovoenergy/comms/model/Feedback.scala]: type signature here, and should probably be a `val`
